File: backend/app/schemas/progress.py
from pydantic import BaseModel


class StrategyInfo(BaseModel):
    key: str
    label: str
    stage: int
    unlock_at: int
    is_unlocked: bool
    is_enabled: bool


class DeckInfo(BaseModel):
    key: str
    label: str
    strategy: str
    description: str
    unlock_at: int
    is_unlocked: bool
    is_enabled: bool
    is_purchasable: bool = False
    shop_price: int | None = None
    card_style: str | None = None


class UserProgressOut(BaseModel):
    unlocked_strategies: list[str]
    enabled_strategies: list[str]
    unlocked_decks: list[str]
    enabled_decks: list[str]
    total_cards_played: int
    strategies: list[StrategyInfo]
    decks: list[DeckInfo]


class UserProgressUpdate(BaseModel):
    enabled_strategies: list[str] | None = None
    enabled_decks: list[str] | None = None


class PurchaseDeckRequest(BaseModel):
    deck_key: str


class PurchaseDeckResponse(BaseModel):
    progress: UserProgressOut
    remaining_capital: float
    remaining_net_worth: float
    purchased_deck_key: str

File: frontend/services/progress.ts
import { api } from "./api";

export interface StrategyInfo {
  key: string;
  label: string;
  stage: number;
  unlock_at: number;
  is_unlocked: boolean;
  is_enabled: boolean;
}

export interface ProgressData {
  unlocked_strategies: string[];
  enabled_strategies: string[];
  total_cards_played: number;
  strategies: StrategyInfo[];
}

export async function getProgress(): Promise<ProgressData> {
  const resp = await api.get<ProgressData>("/api/progress");
  return resp.data;
}

export async function updateProgress(
  enabled_strategies: string[]
): Promise<ProgressData> {
  const resp = await api.patch<ProgressData>("/api/progress", { enabled_strategies });
  return resp.data;
}
